### ISSUE TITLE
ParameterTree, GroupParameter, saveState() and loadState() do not work / do not exist

### ISSUE THREAD

**Issue by jaxankey** [pyqtgraph/pyqtgraph#1130]
Edited to clarify:

`ParameterTree` does not have a means of saving and restoring the collapsed/expanded state of its items. `GroupParameter` and `SimpleParameters` do have `saveState()` and `restoreState()`, but they do not affect the collapsed state of the tree, despite having these keywords in opts; changing the state of the GUI with my mouse (collapsing / expanding items) also does not change the value stored in `opts` or `saveState()`, nor does `setOpts()` or setting `opts` directly affect the collapse state state.

### Code to reproduce

```python
import pyqtgraph

app = pyqtgraph.QtGui.QApplication([])
tree = pyqtgraph.parametertree.ParameterTree()
tree.show()

p1 = pyqtgraph.parametertree.Parameter.create(name='test',  type='bool',  value=True, expanded=True)
p2 = pyqtgraph.parametertree.Parameter.create(name='test2', type='float', value=37.2, expanded=True)

tree.addParameters(p1)
p1.addChild(p2)

p1.setOpts(value    = False) # Works
p1.setOpts(expanded = False) # Does not update the GUI

state = p1.saveState() # Works based on ops, not GUI state
print('Should be True:', state['expanded'] == False)

p1.restoreState(state) # Should collapse the first parameter, but does not. Does update opts.

# Additionally, toggling the collapsed state of the GUI (via mouse, e.g.)
# has no effect on the output of p1.saveState()
```

Basically, 'expanded' seems to be unlinked with the GUI state. Is there a way to control whether a tree item is collapsed?

Thanks in advance! 
Jack

### Tested environment(s)

 * PyQtGraph version: 0.10.0 and 0.11.0rc0
 * Qt Python binding: PyQt5 5.9.2 Qt 5.9.6
 * Python version: 3.7
 * NumPy version: 1.15.4
 * Operating system: Linux Mint 19 & Windows 10
 * Installation method: conda

**Comment by edmund3333** [comment]
maybe related to this issue: https://github.com/pyqtgraph/pyqtgraph/issues/788

**Comment by jaxankey** [comment]
Man. I have been digging through code for hours and cannot seem to figure out how to access the widgets associated with my Parameter objects (to code a solution for myself). I think I'm missing something.

Anyone know how to get basic `collapse()` and `isCollapsed()` functionality starting from a `Parameter` or `ParameterTree` object? Or any other tips for navigating through the tree to accomplish a configuration save / load?

Thanks!

**Comment by 2xB** [comment]
@jaxankey Don't worry, you do not need to code a solution for this yourself. Probably you are helped with the example file "exampels/parametertree.py"? It shows how to react to changed values, see e.g. the "change" function, how to use saveState and restoreState in the "save" and "restore" functions, and also how to change parameters, as seen in "aChanged" and "bChanged" of the custom "ComplexParameter" class.
A good first step for the examples is always to just focus on one feature one is interested in and successively delete all code that is not needed for that feature, until one has a minimal example for that feature.

Does that solve your issues?

**Comment by jaxankey** [comment]
Ack! Sorry. I edited my original post to clarify the issue:

`saveState()`, `loadState()`, `setOpts` and `opts` seem totally unlinked to the collapsed / expanded state of the tree. I cannot for the life of me find a way to affect this, actually, and have not been able to access the underlying Qt widgets (which do have this functionality.

**Comment by 2xB** [comment]
Okay, in that case, might you check if Pull Request #1168 solves the issue?

**Comment by jaxankey** [comment]
Does not seem to. Bear in mind I'm running pyqtgraph v0.10.0 and don't want to break my anaconda install. I did manually overwrite the function in my site-packages directory, but the collapsed / expanded behavior seems to be the same.
________________________________

**Comment by 2xB** [comment]
@jaxankey Okay, in that case, two things:

1. Can you provide a full minimal working example of code that on execution should result in a collapsed tree but does not?

2. If you wish to try out the new version without changing anything in Anaconda: You do not need to install other versions of `pyqtgraph` to try them out. Download the source code of the version you want to try - either via Git or via https://github.com/2xB/pyqtgraph/archive/fix-788.zip in this case. If you download that .zip file, unzip it. Then, start Python in the directory of that downloaded code that contains folders such as `doc`, `examples` or `pyqtgraph`. E.g. you can open the Anaconda Prompt, use cd to move to that directory and execute `python`. From there, if you e.g. `import pyqtgraph as pg`, it will be the `pyqtgraph` version that you just downloaded.

**Comment by jaxankey** [comment]
For #1, I thought I had, but I updated it further to try help clarify (I also had typed `loadState` instead of `restoreState` initially, due to lack of sleep). I will test it on the latest pyqtgraph next.

**Comment by jaxankey** [comment]
Same behavior on v0.11.0rc0. You can try the above code on whatever patched version you have, perhaps?

**Comment by 2xB** [comment]
@jaxankey Thank you very much! I've made a Pull Request that should fix the issue.
Could you check whether Pull Request #1175 also fixes that issue for you? With that, your code should also even work without `saveState` and `restoreState`.

**Comment by jaxankey** [comment]
Thanks, I will test this when I get the chance, but it looks like exactly what was missing.

Is there a way to access the ParameterItem associated with a Parameter? If not, one might gain by providing a handle to it (e.g. Parameter.getParameterItem()).

________________________________

**Comment by 2xB** [comment]
`Parameter`s are unaware of their `ParameterItem`s. The former ones are just containers for the data that is to be presented, the latter ones are the ones presenting that data. `ParameterItem`s learn about changes in the data by listening to various signals from `Parameter`. In computer science, this is called Model-View pattern. Whenever the data (the model) changes, the view updates as quick as possible.
On the other hand, you can have multiple `ParameterItem`s for the same `Parameter`. The example `exampels/parametertree.py` actually shows exactly that, where all parameters are put into two views simultaneously. If you had access to `ParameterItem`s and you used that to change something instead of manipulating the fundamental `Parameter`, you could easily end up in a situation where your data is unexpectedly not synchronized any longer.

So in summary, you should be able to do everything you want on a `Parameter` alone, the `ParameterItem`s are just implementation details. Is there something else broken that currently makes you requiring to access a `ParameterItem`?

**Comment by 2xB** [comment]
I've seen the list you made over at the PR of things that are not synchronized between `Parameter`s and `ParameterItem`s, and I totally get why you wanted access to `ParameterItem`s. I'll take a look at that tomorrow or the day after.

**Comment by jaxankey** [comment]
Naw I don't *need* access, just wanted it so I could hack together my own solution. If the Parameter does everything I'm glad to use it.
________________________________

**Comment by 2xB** [comment]
Added another option to `Parameter`s over at my PR, you might check out if that suits you.
Btw. it turns out: `Parameter`s do indeed currently have a way to access associated `ParameterItem`s. I don't think this will be actively supported, so its use is really only as a quick workaround: A list of items is currently available as attribute `p1.items`.
